source/neuropod/backends/tensorflow/saved_model/loader.cc

```diff
@@ -0,0 +1,329 @@
+    // v2 format will have a variables.index file. The corresponding
+    // variables are stored in the variables.data-?????-of-????? files.
+    const string variables_index_path = io::JoinPath(variables_directory, MetaFilename(kSavedModelVariablesFilename));
+    if (!Env::Default()->FileExists(variables_index_path).ok())
+    {
+        LOG(INFO) << "The specified SavedModel has no variables; no checkpoints "
+                     "were restored. File does not exist: "
+                  << variables_index_path;
+        return Status::OK();
+    }
+    const string variables_path = io::JoinPath(variables_directory, kSavedModelVariablesFilename);
+
+    // Add variables to the graph.
+    Tensor variables_path_tensor(DT_STRING, TensorShape({}));
+    variables_path_tensor.scalar<TFStringType>()() = variables_path;
+
+    std::vector<std::pair<string, Tensor>> inputs = {{string(variable_filename_const_op_name), variables_path_tensor}};
+
+    AddAssetsTensorsToInputs(export_dir, asset_file_defs, &inputs);
+
+    RunMetadata run_metadata;
+    return RunOnce(run_options, inputs, {}, {string(restore_op_name)}, nullptr /* outputs */, &run_metadata, session);
+}
+
+Status LoadSavedModelInternal(const SessionOptions &            session_options,
+                              const RunOptions &                run_options,
+                              const string &                    export_dir,
+                              const std::unordered_set<string> &tags,
+                              SavedModelBundle *const           bundle)
+{
+    TF_RETURN_IF_ERROR(ReadMetaGraphDefFromSavedModel(export_dir, tags, &bundle->meta_graph_def));
+
+    // NEUROPOD MODIFICATION
+    // DebugInfo isn't available in all supported TF versions and is not critical to running a
+    // SavedModel so we ignore it for now
+    //
+    //   TF_RETURN_IF_ERROR(
+    //       ReadSavedModelDebugInfoIfPresent(export_dir, &bundle->debug_info));
+    TF_RETURN_IF_ERROR(LoadMetagraphIntoSession(session_options, bundle->meta_graph_def, &bundle->session));
+    TF_RETURN_IF_ERROR(RestoreSession(run_options, bundle->meta_graph_def, export_dir, &bundle->session));
+    return Status::OK();
+}
+
+} // namespace
+
+SavedModelBundleInterface::~SavedModelBundleInterface() {}
+
+Status LoadMetagraphIntoSession(const SessionOptions &    session_options,
+                                MetaGraphDef &            meta_graph,
+                                std::unique_ptr<Session> *session)
+{
+    Session *session_p = nullptr;
+    TF_RETURN_IF_ERROR(NewSession(session_options, &session_p));
+    session->reset(session_p);
+
+    GraphDef &graph = *meta_graph.mutable_graph_def();
+    TF_RETURN_IF_ERROR(ValidateSavedTensors(graph));
+
+    // NEUROPOD MODIFICATION
+    // This moves the graph to the appropriate device before loading it
+    neuropod::move_graph_to_device(graph, *(session->get()), 0);
```

Comment:
We're creating a confusing dependency model here, by introducing a neuropod namespace dependency on tensorflow. It might be preferable to move the move_graph_to_device logic in the tensorflow namespace, and have the graph loader depend on this logic instead, so the dependency chain remains linear.

Reply:
I'd like to avoid adding things to the `tensorflow` namespace that don't absolutely need to be there. There are some situations where it's necessary (like friend classes that need to be in particular namespaces), but I don't think we need to do that here. Ideally the `tensorflow` namespace would only contain things directly from the TF codebase.

I think this also helps make it clear what's Neuropod code and what came from TF.

I haven't seen anything about enforcing linear dependency chains in C++ namespaces, but it's possible I'm missing something. Do you have a style guide reference for this?

The [Google C++ style guide](https://google.github.io/styleguide/cppguide.html#Namespaces) doesn't mention anything about this as far as I can tell.

Reply:
I understand why you would like to limit modifications to tensorflow, but the circular dependency we're introducing, while not breaking is also not ideal. 

I don't have deep knowledge of either codebase so this suggestion might be naive, but this is essentially what I'm getting at:

> I'm mainly concerned about maintaining what is essentially a copy of Tensorflow code. Is it possible to support these changes in TF instead?

Maybe we can avoid this circular dependency as well as the copy of TF code if the move_graph_to_device API is implemented in TF.

Reply:
Reading these comments in linear order, and just saw your response, and it seems like something you've already considered. Resolving this conversation.